isolatedModules error on global shadowed by imported type

frigus02 opened the pull request:
Fixes #56521

I took inspiration from #56354, which seems very similar.

andrewbranch commented on src/compiler/checker.ts:
```diff
@@ -3603,6 +3604,19 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
+                // here because 'verbatimModuleSyntax' will already have an error for importing a type without 'import type'.
+                if (compilerOptions.isolatedModules && result && isInExternalModule && meaning & SymbolFlags.Value) {
+                    const isGlobal = lookup(globals, name, meaning) === result;
+                    const typeSymbol = isSourceFile(lastLocation!) && lastLocation.locals && lookup(lastLocation.locals, name, SymbolFlags.Type);
```
I don’t think this is quite right—the characteristic of the local symbol that makes it problematic isn’t that it _is a type_; it’s that it _is not a value_. In other words, there could be potential for false positives or false negatives in a few cases here:

- the imported symbol is a type and a value at the same time, like a class (I don’t think this case is possible since `isGlobal` would be false)
- the imported symbol is an uninstantiated namespace
- the imported symbol was unresolvable and so has all `SymbolFlags` (again, shouldn’t be an issue here since `isGlobal` would be false)

Of these corner cases to think about, it sounds like uninstantiated namespaces might pose a problem. Here’s a test:

```ts
// @isolatedModules: true
// @noEmit: true
// @noTypesAndSymbols: true

// @filename: ./types.ts
export namespace Date {
  export type T = any;
}

// @filename: ./bad.ts
import { Date } from './types';
new Date();
```

Maybe a lookup with `~SymbolFlags.Value` instead of `SymbolFlags.Type` would better capture the intent?

frigus02 replied:
Good catch and thanks for the explanation. Changed to `~SymbolFlags.Value` and added the test case.

andrewbranch reviewed (changes requested):
Thanks! I think this is pretty close.

frigus02 commented:
Thanks for the quick review! I think I addressed your comments.

(Not sure about the comment resolving etiquette here. I hope me resolving them is fine?)

frigus02 commented:
The new error currently incorrectly fires for the following Node.js code:

```ts
import { Console } from "node:console";
const c: Console = new Console();
```

That's because `console.Console` is indeed a global. But it's re-exported in the module `node:console`. Importantly, though, the global is `console.Console`, not `Console`, so the import is necessary.

Node.js typings for reference: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/420d2ceb50f728aa715da1250a2a9f27760a640c/types/node/console.d.ts#L412

Weirdly there is no error for this code:

```ts
import { Console } from "node:console";
new Console();
```

I'm going to look into that tomorrow. If you have any ideas, I'd definitely appreciate your help, @andrewbranch.

frigus02 commented:
Aha. I think `SymbolFlags.Value` is not a single bit. I just copied the check from a few lines up. But I think I need to use `(meaning & SymbolFlags.Value) === SymbolFlags.Value` here. I'll test some more tomorrow and update the code and tests.

andrewbranch commented:
Oh, good catch. Yeah, I think something like that should work, or alternatively a lookup with `SymbolFlags.Type | SymbolFlags.NamespaceModule`

frigus02 commented:
As far as I can tell this looks good now. I did not use `SymbolFlags.Type | SymbolFlags.NamespaceModule`, yet. But happy to switch to that if you prefer.

andrewbranch commented:
@typescript-bot perf test this
@typescript-bot test top200

andrewbranch reviewed (approved):
Thanks @frigus02!
